### PR TITLE
Stabilize Digit Party DQN and document the 3x3 result

### PR DESCRIPTION
**Pull request by zhjngli**
## Summary
- stabilize Digit Party DQN by masking invalid actions during training and bootstrapping only over legal next moves
- give the DQN path its own simpler network/update setup and retune the training schedule for frequent small replay updates over a much longer run
- add evaluation, best-checkpoint saving, unique-state tracking, and README/result artifacts documenting the successful 3x3 run

## Testing
- `pipenv run python -m py_compile src/cli.py src/games/digit_party/run_helpers.py src/games/digit_party/train_q_deep.py`

### CODE REVIEW

File: src/learners/deep_q.py [hunk 1]
```diff
@@ -110,23 +138,11 @@ def run_game_once(self) -> None:
             self.steps += 1
 
             score = self.game.calculate_reward(state)
+            state_snapshot = copy.deepcopy(state)
             # print(f"state:\n{state.board}")
             # print(f"next: {state.next}")  # type: ignore
 
-            # epsilon greedy
```
Comment (zhjngli) [inline src/learners/deep_q.py:116]: let's keep the note about epsilon greedy here. not sure if the choose_action and valid_actions methods are needed to be extracted but it's ok for now

File: src/learners/deep_q.py [hunk 2]
```diff
@@ -128,14 +147,32 @@ def train(self) -> None:
             if i % self.episodes_per_memory_save == 0:
                 self.save_memory(f"ep_{i:07d}_memory.pkl")
 
-            # TODO: track efficacy of learning (e.g. play some number of games and track score)
```
Comment (zhjngli) [inline src/learners/deep_q.py:103]: the todo is useful to keep around

File: src/games/digit_party/train_q_deep.py [hunk 3]
```diff
@@ -13,14 +14,26 @@
...
+    conv_filters=opt_nn_params.conv_filters,
+    dense_layers=opt_nn_params.dense_layers,
+    dense_units=opt_nn_params.dense_units,
+    learning_rate=opt_nn_params.learning_rate,
+    batch_size=opt_nn_params.batch_size,
+    epochs=1,
+    dropout_rate=0.0,
+    output_activation=opt_nn_params.output_activation,
+)
+
```
Comment (zhjngli) [inline src/games/digit_party/train_q_deep.py:37]: i'm confused what this is doing since it looks like it's just using the opt_nn_params?
